File: src/main/java/xf/xflp/base/XFLPModel.java
package xf.xflp.base;

import xf.xflp.base.container.Container;
import xf.xflp.base.item.Item;
import xf.xflp.base.monitor.StatusManager;


/** 
 * Copyright (c) 2012-2022 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 *
 * The XFLPModel holds all necessary input data. The
 * nodeArr contains all read nodes with their attitudes.
 * The metric can only be reached through this model for
 * getting distances or times between nodes. 
 * @author hschneid
 */
public class XFLPModel {

	protected Item[] items;

	/* Only one object per container type, solution will copy these objects */
	protected final Container[] containerTypes;

	protected final XFLPParameter parameter;
	private StatusManager statusManager;

	/* Result objects */
	private Container[] containers = new Container[0];
	private Item[] unplannedItems = new Item[0];

	/**
	 * Initialize an optimization model object with the given input data. It contains the general
	 * parameter for all optimization procedures. It holds no solution information.
	 */
	public XFLPModel(Item[] items, Container[] containerTypes, XFLPParameter parameter, StatusManager statusManager) {
		this.items = items;
		this.containerTypes = containerTypes;
		this.parameter = parameter;
		this.statusManager = statusManager;
	}

	public Container[] getContainerTypes() {
		return containerTypes;
	}

	public XFLPParameter getParameter() {
		return parameter;
	}

	public Item[] getItems() {
		return items;
	}

	/**
	 * @return the containers
	 */
	public final Container[] getContainers() {
		return containers;
	}

	/**
	 * @param containers the containers to set
	 */
	public final void setContainers(Container[] containers) {
		this.containers = containers;
	}

	/**
	 * @return the unplannedItems
	 */
	public final Item[] getUnplannedItems() {
		return unplannedItems;
	}

	/**
	 * @param unplannedItems the unplannedItems to set
	 */
	public final void setUnplannedItems(Item[] unplannedItems) {
		this.unplannedItems = unplannedItems;
	}

	public void setItems(Item[] items) {
		this.items = items;
	}

	public StatusManager getStatusManager() {
		return statusManager;
	}
}

File: src/xflp/base/xflp_model.py
"""
Copyright (c) 2012-2025 Holger Schneider
All rights reserved.

This source code is licensed under the MIT License (MIT) found in the
LICENSE file in the root directory of this source tree.

The XFLPModel holds all necessary input data.
"""

from typing import List, TYPE_CHECKING

if TYPE_CHECKING:
    from xflp.base.container import Container
    from xflp.base.item import Item
    from xflp.base.monitor import StatusManager
    from .xflp_parameter import XFLPParameter


class XFLPModel:
    """
    The XFLPModel holds all necessary input data for the optimization.
    """

    def __init__(self, items: List['Item'], container_types: List['Container'],
                 parameter: 'XFLPParameter', status_manager: 'StatusManager'):
        """
        Initialize an optimization model object with the given input data.

        Args:
            items: Array of items to be packed
            container_types: Array of container types (only one per type, will be copied)
            parameter: General parameters for all optimization procedures
            status_manager: Manager for status monitoring
        """
        self._items = items
        self._container_types = container_types
        self._parameter = parameter
        self._status_manager = status_manager

        # Result objects
        self._containers: List['Container'] = []
        self._unplanned_items: List['Item'] = []

    def get_container_types(self) -> List['Container']:
        """Get the container types."""
        return self._container_types

    def get_parameter(self) -> 'XFLPParameter':
        """Get the parameters."""
        return self._parameter

    def get_items(self) -> List['Item']:
        """Get the items."""
        return self._items

    def get_containers(self) -> List['Container']:
        """Get the result containers."""
        return self._containers

    def set_containers(self, containers: List['Container']) -> None:
        """Set the result containers."""
        self._containers = containers

    def get_unplanned_items(self) -> List['Item']:
        """Get the unplanned items."""
        return self._unplanned_items

    def set_unplanned_items(self, unplanned_items: List['Item']) -> None:
        """Set the unplanned items."""
        self._unplanned_items = unplanned_items

    def set_items(self, items: List['Item']) -> None:
        """Set the items."""
        self._items = items

    def get_status_manager(self) -> 'StatusManager':
        """Get the status manager."""
        return self._status_manager
